### PR TITLE
planner: add 2pb logic and fix resolve index bug for partitionTopN

### PR DESCRIPTION
**Pull request by LittleFall**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/39792

Problem Summary: partitionTopN lack of 2pb logic and have a bug about resolve index.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by winoros — APPROVED** [review]

**Review by AilinKid — APPROVED** [review]

File: planner/core/logical_plans.go [hunk 1]
```diff
@@ -163,6 +163,7 @@ type LogicalJoin struct {
 	rightPreferJoinType uint
 
 	EqualConditions []*expression.ScalarFunction
+	// NAEQConditions means null aware equal conditions, which is used for null aware anti joins.
```
Comment (gengliqi) [inline planner/core/logical_plans.go:166]:
```suggestion
	// NAEQConditions means null aware equal conditions, which is used for null aware semi joins.
```

**Comment by windtalker** [comment]
close as https://github.com/pingcap/tidb/pull/42334 include all the fix in this pr

**Pull request closed without merging**